src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,587 @@
+//!
+//! ```
+//! use hdrsample::serialization::interval_log;
+//!
+//! let mut log = Vec::new();
+//! log.extend_from_slice(b"#I'm a comment\n");
+//! log.extend_from_slice(b"Tag=a,0.123,1.007,2.769,base64EncodedHisto\n");
+//! log.extend_from_slice(b"1.456,1.007,2.769,base64EncodedHisto\n");
+//! log.extend_from_slice(b"3.789,1.007,2.769,base64EncodedHisto\n");
+//! log.extend_from_slice(b"Tag=b,4.123,1.007,2.769,base64EncodedHisto\n");
+//! log.extend_from_slice(b"5.456,1.007,2.769,base64EncodedHisto\n");
+//! log.extend_from_slice(b"#Another comment\n");
+//!
+//! let iter = interval_log::IntervalLogIterator::new(&log);
+//!
+//! let count = iter.map(|r| r.unwrap())
+//!     // only look at intervals (which are the only non-comment lines in this log)
+//!     .filter_map(|e| match e {
+//!         interval_log::LogEntry::Interval(ilh) => Some(ilh),
+//!          _ => None
+//!     })
+//!     // do any filtering you want
+//!     .filter(|ilh| ilh.start_timestamp() >= 3.0)
+//!     .count();
+//!
+//! assert_eq!(3, count);
+//! ```
+//!
+//! Write a log.
+//!
+//! ```
+//! use std::str;
+//! use hdrsample;
+//! use hdrsample::serialization;
+//! use hdrsample::serialization::interval_log;
+//!
+//! let mut buf = Vec::new();
+//! let mut serializer = serialization::V2Serializer::new();
+//!
+//! let mut h = hdrsample::Histogram::<u64>::new_with_bounds(
+//!     1, u64::max_value(), 3).unwrap();
+//! h.record(12345).unwrap();
+//!
+//! // limit scope of mutable borrow of `buf`
+//! {
+//!     let mut header_writer = interval_log::IntervalLogHeaderWriter::new(
+//!         &mut buf, &mut serializer);
+//!     header_writer.write_comment("Comments are great").unwrap();
+//!     header_writer.write_start_time(123.456789).unwrap();
+//!
+//!     let mut log_writer = header_writer.into_log_writer();
+//!
+//!     log_writer.write_comment(
+//!         "You can have comments anywhere in the log").unwrap();
+//!
+//!     log_writer
+//!         .write_histogram(
+//!             &h,
+//!             1.234,
+//!             5.678,
+//!             interval_log::Tag::new("im-a-tag"),
+//!             1.0)
+//!         .unwrap();
+//! }
+//!
+//! // `buf` is now full of stuff; we check for the first line
+//! assert_eq!("#Comments are great\n", &str::from_utf8(&buf).unwrap()[0..20]);
+//! ```
+
+extern crate base64;
+
+use std::{fmt, io, ops, str};
+use std::fmt::Write;
+
+use nom::{double, line_ending, not_line_ending, IResult};
+
+use super::super::{Counter, Histogram};
+use super::Serializer;
+
+/// Write headers for an interval log.
+///
+/// This type only allows writing comments and headers. Once you're done writing those things, use
+/// `into_log_writer()` to convert this into an `IntervalLogWriter`.
+pub struct IntervalLogHeaderWriter<'a, 'b, W: 'a + io::Write, S: 'b + Serializer> {
+    internal_writer: InternalLogWriter<'a, 'b, W, S>,
+}
+
+impl<'a, 'b, W: 'a + io::Write, S: 'b + Serializer> IntervalLogHeaderWriter<'a, 'b, W, S> {
```

Comment:
I don't think the `IntervalLogHeaderWriter` should be parameterized on `S`. Instead, the call to `into_log_writer` should be. The header writes are all independent of the `Serializer` as far as I can tell?